pytest_mh/utils/firewall.py

```diff
@@ -0,0 +1,302 @@
+        """
+        self.__add_action(what, action="reject")
+
+    def drop(self, what: ServiceSpec | list[ServiceSpec]) -> None:  # type: ignore[override]
+        """
+        Drop incoming traffic on given port or service.
+
+        .. code-block:: python
+
+            firewall.drop(389)  # missing protocol defaults to "tcp"
+            firewall.drop((389, "tcp"))
+            firewall.drop("ldap")
+            firewall.drop(["ldap", "ldaps"])
+
+
+        :param what: Port (integer), (port, protocol) tuple or service (string).
+        :type what: ServiceSpec | list[ServiceSpec]
+        """
+        self.__add_action(what, action="drop")
+
+    def add_rich_rule(self, rule: str) -> None:
+        """
+        Add rich rule.
+
+        :param rule: Firewalld rich rule.
+        :type rule: str
+        """
+        self.logger.info(f'Firewalld: adding rich rule "{rule}" on {self.host.hostname}')
+        self.host.ssh.exec(["firewall-cmd", "--add-rich-rule", rule], log_level=SSHLog.Error)
+
+    def remove_rich_rule(self, rule: str) -> None:
+        """
+        Remove rich rule.
+
+        :param rule: Firewalld rich rule.
+        :type rule: str
+        """
+        self.logger.info(f'Firewalld: removing rich rule  "{rule}" on {self.host.hostname}')
+        self.host.ssh.exec(["firewall-cmd", "--remove-rich-rule", rule], log_level=SSHLog.Error)
+
+    def __add_action(
+        self, what: ServiceSpec | list[ServiceSpec], *, action: Literal["accept", "reject", "drop"]
+    ) -> None:
+        def __add_port(port: int, protocol: str):
+            self.add_rich_rule(f'rule priority="{self._priority}" port port="{port}" protocol="{protocol}" {action}')
+
+        def __add_service(name: str):
+            self.add_rich_rule(f'rule priority="{self._priority}" service name="{name}" {action}')
+
+        if not isinstance(what, list):
+            what = [what]
+
+        for item in what:
+            if isinstance(item, str):
+                __add_service(item)
+                continue
+
+            port, protocol = self.parse_port_spec(item)
+            __add_port(port, protocol)
+
+        self._priority -= 1
+
+
+class WindowsFirewall(GenericFirewall):
+    """
+    Configure Windows firewall.
+
+    All changes are automatically reverted when a test is finished.
+    """
+
+    def __init__(self, host: MultihostHost) -> None:
+        """
+        :param host: Remote host instance.
+        :type host: MultihostHost
+        """
+        super().__init__(host)
+        self._changed: bool = False
+        """Did we change anything?"""
+
+        self._rules: list[str] = []
+        self._backup: str = "C:\\.mh_firewall.bak.wfw"
+
+    def setup_when_used(self):
+        """
+        Create a backup of current firewall configuration.
+
+        :meta private:
+        """
+        super().setup_when_used()
+        self.logger.info(f"Windows Firewall: creating backup on {self.host.hostname}")
```

Comment:
Worth adding the backup path + filename to this log message?

Reply:
Done.